src/PrestaShopBundle/Entity/Repository/CmsPageRepository.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShopBundle\Entity\Repository;
+
+use CMS;
+use Language;
+use PrestaShop\PrestaShop\Core\Domain\CmsPage\CmsPageSettings;
+use PrestaShop\PrestaShop\Core\Domain\CmsPage\Exception\CannotDuplicateCmsPageException;
+use PrestaShop\PrestaShop\Core\Domain\CmsPage\Exception\CmsPageNotFoundException;
+use PrestaShop\PrestaShop\Core\Domain\CmsPage\ValueObject\CmsPageId;
+use PrestaShop\PrestaShop\Core\Repository\AbstractObjectModelRepository;
+use PrestaShop\PrestaShop\Core\Util\Entity\EntityNameDuplicator;
+
+class CmsPageRepository extends AbstractObjectModelRepository
+{
+    /**
+     * @var EntityNameDuplicator
+     */
+    private $entityNameDuplicator;
+
+    /**
+     * @param EntityNameDuplicator $entityNameDuplicator
+     */
+    public function __construct(
+      EntityNameDuplicator $entityNameDuplicator
+  ) {
+      $this->entityNameDuplicator = $entityNameDuplicator;
+  }
+
+    /**
+     * Duplicates a CMS page
+     *
+     * @param CmsPageId $cmsPageId
+     *
+     * @return CmsPageId
+     *
+     * @throws CmsPageNotFoundException
+     * @throws CannotDuplicateCmsPageException
+     */
+    public function duplicate(CmsPageId $cmsPageId): CmsPageId
+    {
+        /** @var Cms $cms */
+        $cms = $this->getObjectModel(
+          $cmsPageId->getValue(),
+          Cms::class,
+          CmsPageNotFoundException::class
+        );
+
+        // Reset object ID, make it inactive and save it
+        $cms->id = null;
+        $cms->meta_title = $this->entityNameDuplicator->getNewLocalizedNames(
```

Comment:
entity name duplicator is a nice service, but what I am still in doubt (as we discussed privately), that this duplicator service doesn't really fit in the repository pattern.
If we keep following product example, there should be a CmsPageDuplicator service responsible for changing the properties and then repository basically only saves it (thats what its for - persistence).

Though I feel Im going to far with my requests and I even think that I made it wrong in product repository too. Instead of having a redundant "duplicate" method in repository, product Id must have been unset in the duplicator and then common ProductRepository::add() method should have been called.
I need help @jolelievre ! could you help me decide what we do with this PR, because it really looks good and clean, just these little details left (which are kind looping around cuz of my mistakes)

Reply:
btw the validation is missing :sweat_smile:

Reply:
I think it's ok to have the duplicate method here, for product it's a bit different because the domain is more complex it's not only about duplicating an entity it's also all its relations, and the repository is already quite big, so it makes sense to divide it into several services

Reply:
ok, glad to hear. Though we are missing the validation for this cms creation. Not sure if its a big deal, just if some error ever happens, the legacy PrestashopException will bubble out to the surface :zany_face: 

Im not requesting this, but just to imagine what I thought would have been ideal (before I forget :smile:):
```
class CmsPageRepository
{
    ...
    // this coud have been reused in addCmsPageHandler
    public function add(CmsPage $cmsPage): CmsPage
    {
        // as in other recent pages, validates properties one by one and ensures no legacy exception bubbles out
        $this->cmsPageValidator->validate($cmsPage);
        $this->addObjectModel($cmsPage...);
        
        return $cmsPage;
    }
 }   
    
class CmsPageDuplicator
{
    ...
    public function duplicate(CmsPageId $cmsPageId): CmsPage
    {
        $cmsPage = $this->cmsPageRepository->get($cmsPageId);
        $cmsPage->id = null;
        $cmsPage->name = $this->nameDuplicator->getNewLocalizedName(...);
        $cmsPage->{some other changes if needed)...;
        
        return $this->cmsPageRepository->add($cmsPage);
    }
}
```
So what we would get from that:
1. all props are well validated
2. CmsPageRepository has a reusable "add" method
3. Responsibilities are clear - duplicator duplicates, repository persists (so we don't have additional logic inject into repository).
Negative stuff:
1. a little more work to do (especially knowing that all those repository services ain't designed for long term, as they are still relying on object model)